### PR TITLE
Two-Pass Iterative SRF: sequential boosting of surface corrections

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

The Surface Refinement Head (SRF) corrects surface predictions in a single MLP pass. Like a single weak learner in an ensemble, it captures the dominant correction mode but systematically misses second-order residuals.

**Adding a second, smaller SRF pass** applies gradient-boosting logic to the correction pathway: SRF2 operates on the *already-corrected* prediction from SRF1, learning what SRF1 consistently gets wrong. This is the same principle as XGBoost/gradient boosting — each corrector focuses on the errors left by the previous one.

**Why this is different from wider SRF heads (PR #2252, closed):**
- Wider SRF (192→384) added more capacity to a single pass → overfitting, all metrics worse
- Two-pass adds a separate corrector operating on SRF1's residuals → different error modes, much smaller parameter budget (~28K vs 261K new params for wider heads)
- Key insight from #2252: 192-dim is well-matched to the data signal. The bottleneck is the *structure* of correction (single pass), not raw capacity

**Why the `.detach()` matters:** SRF2 receives `(base_pred + correction1).detach()` — gradients do NOT flow back through SRF1 via SRF2's loss. Each pass is an independent corrector, exactly like gradient boosting's stage-wise fitting.

**Physical motivation:** Surface pressure has multiple error sources at different scales:
1. Slow global shifts from backbone bias at a given Re → SRF1 handles this
2. Sharp local features at leading/trailing edge (suction peaks) → SRF2 learns the second-order residual
3. Inter-foil coupling errors in tandem → residuals SRF1 misses, SRF2 can catch

**Zero-init safety:** SRF2's output layer is zero-initialized — at epoch 0, `correction2 = 0` and the model is identical to the single-pass baseline. No regression risk.

**Parameter overhead:** SRF2 at `hidden_dim=96, n_layers=2`: ≈28,000 params. Negligible. VRAM: ~+30 MB.

**Expected impact:** -2 to -4% p_tan, -2 to -3% p_oodc.

**Confidence:** MEDIUM. Gradient boosting on residuals is a proven ML paradigm. Zero-init safety is high.

---

## Instructions

Add a two-pass sequential correction pathway to the surface refinement head (~30 lines in `train.py`).

### Step 1: Add config flags

Near the other `surface_refine` flags (search for `surface_refine_hidden`), add:

```python
surface_refine2: bool = False
surface_refine2_hidden: int = 96
surface_refine2_layers: int = 2
```

### Step 2: Instantiate the second SRF head

Find where `refine_head = SurfaceRefinementHead(...)` is created. Right after it, add:

```python
if cfg.surface_refine2:
    refine_head2 = SurfaceRefinementHead(
        n_hidden=cfg.n_hidden,
        out_dim=3,
        hidden_dim=cfg.surface_refine2_hidden,
        n_layers=cfg.surface_refine2_layers
    )
    # Zero-init output layer → correction2 = 0 at epoch 0 (identical to baseline)
    nn.init.zeros_(refine_head2.layers[-1].weight)
    nn.init.zeros_(refine_head2.layers[-1].bias)
```

### Step 3: Apply second pass in forward

Find where SRF correction is applied:
```python
pred_surf = base_pred_surf + refine_head(hidden_surf, base_pred_surf)
```

Change to:
```python
correction1 = refine_head(hidden_surf, base_pred_surf)
pred_surf = base_pred_surf + correction1
if cfg.surface_refine2:
    correction2 = refine_head2(hidden_surf, pred_surf.detach())
    pred_surf = pred_surf + correction2
```

**The `.detach()` is critical** — prevents SRF2 gradients from flowing back through SRF1, keeping each corrector independent.

### Step 4: Checkpoint save/load

Find where `refine_head` state dict is saved/loaded and add analogous handling for `refine_head2` when `cfg.surface_refine2` is True.

### Step 5: Run 2 seeds

```bash
# Seed 42
cd cfd_tandemfoil && python train.py \
  --agent fern --wandb_name "fern/srf-two-pass-s42" \
  --wandb_group "round21/srf-two-pass" \
  --seed 42 \
  --surface_refine2 --surface_refine2_hidden 96 --surface_refine2_layers 2 \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aft_foil_srf_hidden 192 --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame --wake_deficit_feature

# Seed 73: identical with --seed 73 --wandb_name "fern/srf-two-pass-s73"
```

New flags vs baseline: `--surface_refine2 --surface_refine2_hidden 96 --surface_refine2_layers 2`

### Step 6: Report results

Table: p_in, p_oodc, p_tan, p_re for both seeds + 2-seed average, vs baseline. Include W&B run IDs. Note any epoch count difference vs baseline (~1% compute overhead expected — negligible).

---

## Baseline

| Metric | 2-seed avg | Target to beat |
|--------|-----------|----------------|
| p_in   | **11.979** | < 11.98 |
| p_oodc | **7.643**  | < 7.65  |
| p_tan  | **28.341** | < 28.34 |
| p_re   | **6.300**  | < 6.30  |

Baseline W&B: `hgml7i2r` (seed 42), `qic03vrg` (seed 73) — PR #2213.

**Reproduce baseline:**
```bash
cd cfd_tandemfoil && python train.py --agent fern --wandb_name "fern/baseline" \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame --wake_deficit_feature
```

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT fern:

## Results

### Two-Pass Iterative SRF (gradient boosting on surface corrections)

| Metric | Seed 42 | Seed 73 | 2-seed avg | Baseline | Δ% |
|--------|---------|---------|------------|----------|----|
| **p_in** | 12.543 | 12.096 | **12.319** | 11.979 | **+2.8%** ✗ |
| **p_oodc** | 8.369 | 7.523 | **7.946** | 7.643 | **+4.0%** ✗ |
| **p_tan** | 28.829 | 28.053 | **28.441** | 28.341 | **+0.4%** ✗ |
| **p_re** | 6.672 | 6.382 | **6.527** | 6.300 | **+3.6%** ✗ |

**val_loss (best):** s42=0.383, s73=0.376

**W&B runs:** `xgzt2sz0` (s42), `6hzyktgn` (s73)

**Group:** `round21/srf-two-pass`

**SRF2 params:** 28,803 (hidden=96, layers=2)

**Commands used:**
```bash
cd cfd_tandemfoil && CUDA_VISIBLE_DEVICES=0 python train.py \
  --agent fern --wandb_name "fern/srf-two-pass-s42" \
  --wandb_group "round21/srf-two-pass" \
  --seed 42 \
  --surface_refine2 --surface_refine2_hidden 96 --surface_refine2_layers 2 \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aft_foil_srf_hidden 192 --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame --wake_deficit_feature
# Seed 73 identical with --seed 73 --wandb_name "fern/srf-two-pass-s73"
```

**Training:** 147 epochs before 180-min timeout. Both seeds best at epoch 146. Slightly fewer epochs than baseline (149) due to the extra SRF2 forward pass.

**Peak memory:** ~62 GB (GPU 0), ~64 GB (GPU 1) — marginally higher than baseline due to SRF2.

### What happened

**Negative result.** The two-pass sequential SRF degraded all metrics, with substantial seed variance:
- **p_in +2.8%** — significant regression
- **p_oodc +4.0%** — worst regression, also massive seed gap (s42=8.369 vs s73=7.523)
- **p_tan +0.4%** — marginal, but seed 73 actually beat baseline (28.053 < 28.341) while seed 42 was worse
- **p_re +3.6%** — significant regression

Despite the `.detach()` on SRF2 input preventing gradient flow back through SRF1, the two-pass approach appears to interfere with training dynamics. Possible reasons:

1. **Loss signal confusion:** The loss sees the combined output of SRF1+SRF2, but SRF1 only gets gradients from its own corrections (not SRF2's). SRF1 may converge to a suboptimal correction since it doesn't "know" SRF2 will fix its residuals — but SRF2's tiny capacity (28K params) can't compensate.

2. **Optimization interference:** Even with detach, having two heads on the surface pathway changes the effective learning dynamics. SRF2 adds noise to the training signal that SRF1 receives through the shared loss.

3. **Similar to wider SRF (#2252):** High seed variance (especially p_oodc) suggests the added complexity makes optimization less stable, consistent with the wider SRF finding.

**Key insight:** The single-pass SRF with 192-dim is already well-optimized for this task. All attempts to modify the surface refinement pathway — wider heads, two-pass boosting — have degraded results. The SRF architecture appears to be at a local optimum.

### Suggested follow-ups
- **Close this line of investigation** — SRF modifications consistently hurt. The 3-layer MLP at hidden=192 is the right capacity.
- Focus improvement efforts elsewhere: backbone architecture, loss design, or data augmentation rather than SRF modifications.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2259 because ADVISOR: **Closing — negative result across all metrics.**

| Metric | 2-seed avg | Baseline | Δ% |
|--------|-----------|----------|-----|
| p_in | 12.319 | 11.979 | +2.8% |
| p_oodc | 7.946 | 7.643 | +4.0% |
| p_tan | 28.441 | 28.341 | +0.4% |
| p_re | 6.527 | 6.300 | +3.6% |

**Analysis:** The two-pass SRF degraded all four metrics despite the careful gradient-boosting design with `.detach()` and zero-init. The core finding is consistent with PR #2252 (wider SRF): the single-pass 3-layer 192-dim SRF is already well-matched to the data's correction structure. Adding a second corrector — whether wider (384-dim) or sequential (two-pass boosting) — introduces optimization instability (note the massive p_oodc seed variance: 8.369 vs 7.523).

**Conclusion:** SRF modifications are exhausted as a research direction. The surface refinement pathway is at a local optimum. Future improvements should target the backbone, loss function, or input representations rather than SRF architecture.

Good experimental work, fern — clean implementation and insightful analysis. Your observation that all SRF mods consistently hurt is a valuable negative result that sharpens our research strategy..
